railties/lib/rails/generators/generated_attribute.rb

```diff
@@ -25,6 +43,10 @@ def parse(column_definition)
           type, attr_options = *parse_type_and_options(type)
           type = type.to_sym if type
 
+          if type && !TYPE_OPTIONS.include?(type.to_s)
+            raise Thor::Error, "Unsupported type '#{type}' for attribute '#{name}'."
+          end
```

Comment:
I think this might be too restrictive.  For example, it doesn't allow any of these [Postgres column types](https://github.com/rails/rails/blob/e803ab6e80e2fc0b772e2135ce85111cdce23e2f/activerecord/lib/active_record/connection_adapters/postgresql/schema_definitions.rb#L177-L180).

One possibility to partially address the [original issue](https://discuss.rubyonrails.org/t/why-do-generators-let-me-make-mistakes-on-column-types/75493) would be to check if `type` is not a recognized type *but* `name` is.  But that may still be too restrictive, e.g. `text:citext` should be an accepted name:type combination.

Reply:
I was thinking this too, but then again I also wonder how many people are using generators for these complex PG types?  Maybe it's good enough?  Or maybe we could add a `--i-know-what-i'm-doing` flag and ignore any error 😬

Reply:
There is a `valid_type?` method defined on `AbstractAdapter` which is used by the schema_dumper: https://github.com/rails/rails/blob/e7317a253819b917a3e2284e53347c67246a934e/activerecord/lib/active_record/schema_dumper.rb#L149

It looks at `NATIVE_DATABASE_TYPES` defined in each adapter.
https://github.com/rails/rails/blob/817389d1dd4fd531a8df58e4f10d9da71147d193/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L64

Reply:
I've updated the implementation to look at the adapters supported native types if the type isn't in the `DEFAULT_TYPES`. So it works for example with Sqlites `json` types:
<img width="837" alt="image" src="https://user-images.githubusercontent.com/28561/83803360-b20cf200-a6ac-11ea-9488-ada494fb3e02.png">

Reply:
If this gets merged I can look at adding DidYouMean suggestions to suggest the inverse if it's valid
> Unknown type 'name' for attribute 'string'.
  Did you mean?  name:string

Reply:
I could add a `--skip-type-check` option just like `--skip-collosion-check`

Reply:
or maybe add `DidYouMean` in this one itself?